include/dgl/immutable_graph.h

```diff
@@ -17,317 +17,444 @@
-    typedef std::shared_ptr<CSR> Ptr;
-
-    /*
-     * This vector provides interfaces similar to std::vector.
-     * The main difference is that the memory used by the vector can be allocated
-     * outside the vector. The main use case is that the vector can use the shared
-     * memory that is created by another process. In this way, we can access the
-     * graph structure loaded in another process.
-     */
-    template<class T>
-    class vector {
-     public:
-      vector() {
-        this->arr = nullptr;
-        this->capacity = 0;
-        this->curr = 0;
-        this->own = false;
-      }
+  // Create a csr graph that has the given number of verts and edges.
+  CSR(int64_t num_vertices, int64_t num_edges);
+  // Create a csr graph that shares the given indptr and indices.
+  CSR(IdArray indptr, IdArray indices, IdArray edge_ids);
+  // Create a csr graph whose memory is stored in the shared memory
+  //   and the structure is given by the indptr and indcies.
+  CSR(IdArray indptr, IdArray indices, IdArray edge_ids,
+      const std::string &shared_mem_name);
+  // Create a csr graph whose memory is stored in the shared memory
+  //   that has the given number of verts and edges.
+  CSR(const std::string &shared_mem_name, int64_t num_vertices, int64_t num_edges);
+
+  void AddVertices(uint64_t num_vertices) override {
+    LOG(FATAL) << "CSR graph does not allow mutation.";
+  }
 
-      /*
-       * Create a vector whose memory is allocated outside.
-       * Here there are no elements in the vector.
-       */
-      vector(T *arr, size_t size) {
-        this->arr = arr;
-        this->capacity = size;
-        this->curr = 0;
-        this->own = false;
-      }
+  void AddEdge(dgl_id_t src, dgl_id_t dst) override {
+    LOG(FATAL) << "CSR graph does not allow mutation.";
+  }
 
-      /*
-       * Create a vector whose memory is allocated by the vector.
-       * Here there are no elements in the vector.
-       */
-      explicit vector(size_t size) {
-        this->arr = static_cast<T *>(malloc(size * sizeof(T)));
-        this->capacity = size;
-        this->curr = 0;
-        this->own = true;
-      }
+  void AddEdges(IdArray src_ids, IdArray dst_ids) override {
+    LOG(FATAL) << "CSR graph does not allow mutation.";
+  }
 
-      ~vector() {
-        // If the memory is allocated by the vector, it should be free'd.
-        if (this->own) {
-          free(this->arr);
-        }
-      }
+  void Clear() override {
+    LOG(FATAL) << "CSR graph does not allow mutation.";
+  }
 
-      vector(const vector &other) = delete;
-
-      /*
-       * Initialize the vector whose memory is allocated outside.
-       * There are no elements in the vector.
-       */
-      void init(T *arr, size_t size) {
-        CHECK(this->arr == nullptr);
-        this->arr = arr;
-        this->capacity = size;
-        this->curr = 0;
-        this->own = false;
-      }
+  bool IsMultigraph() const override {
+    return false;
+  }
```

Comment:
How about this. I will change the member to `Lazy<bool> is_multigraph_`. User has an option to not specify whether the graph is a multigraph or not during the construction. When `IsMultigraph()` is called, it will try `Get` the value by a scan of the full graph. Since the graph is immutable, the scan should only be done once. (in fact, for mutable graph, we probably should also do this as a cache).

Reply:
It seems IsMultigraph() isn't used. People only call ImmutableGraph::IsMultigraph(). I guess it doesn't really matter that much what value is returned from IsMultigraph().

Reply:
Immutable graph should return `AnyGraph()->IsMultigraph()` technically. My current implementation is not correct.

Reply:
Done. See if works for you.